### Feature Engineering

In [12]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline
train = pd.read_csv('train_users_2.csv')
#train = train.sample(frac=0.1)
sessions = pd.read_csv('sessions.csv')
groupby = sessions.groupby(by='user_id', sort=False)

In [13]:
train

id date_account_created  timestamp_first_active  \
0       gxn3p5htnn           2010-06-28          20090319043255   
1       820tgsjxq7           2011-05-25          20090523174809   
2       4ft3gnwmtx           2010-09-28          20090609231247   
3       bjjt8pjhuk           2011-12-05          20091031060129   
4       87mebub9p4           2010-09-14          20091208061105   
5       osr2jwljor           2010-01-01          20100101215619   
6       lsw9q7uk0j           2010-01-02          20100102012558   
7       0d01nltbrs           2010-01-03          20100103191905   
8       a1vcnhxeij           2010-01-04          20100104004211   
9       6uh8zyj2gn           2010-01-04          20100104023758   
10      yuuqmid2rp           2010-01-04          20100104194251   
11      om1ss59ys8           2010-01-05          20100105051812   
12      k6np330cm1           2010-01-05          20100105060859   
13      dy3rgx56cu           2010-01-05          20100105083259   
14      ju3h98ch3w           2010-01-07          20100107055820   
15      v4d5rl22px           2010-01-07          20100107204555   
16      2dwbwkx056           2010-01-07          20100107215125   
17      frhre329au           2010-01-07          20100107224625   
18      cxlg85pg1r           2010-01-08          20100108015641   
19      gdka1q5ktd           2010-01-10          20100110010817   
20      qdubonn3uk           2010-01-10          20100110152120   
21      qsibmuz9sx           2010-01-10          20100110220941   
22      80f7dwscrn           2010-01-11          20100111031438   
23      jha93x042q           2010-01-11          20100111224015   
24      7i49vnuav6           2010-01-11          20100111230808   
25      al8bcetz0g           2010-01-12          20100112131444   
26      bjg0m5otl3           2010-01-12          20100112155420   
27      hfrl5gle36           2010-01-12          20100112205949   
28      tp6x3md0n4           2010-01-13          20100113044650   
29      hql77nu2lk           2010-01-13          20100113064333   
...            ...                  ...                     ...   
213421  c98s3h7kgj           2014-06-30          20140630231137   
213422  ytmpiwb8hj           2014-06-30          20140630231246   
213423  3dx1jk6yk2           2014-06-30          20140630231548   
213424  hcfj07iowv           2014-06-30          20140630231859   
213425  l1f71f9vsj           2014-06-30          20140630232119   
213426  15bj4ahmhf           2014-06-30          20140630232331   
213427  qwpybxfjdl           2014-06-30          20140630232539   
213428  k4t61wuvyq           2014-06-30          20140630232634   
213429  mhh7b52z44           2014-06-30          20140630232712   
213430  79wk7k2k5t           2014-06-30          20140630233132   
213431  ftwmocvwlq           2014-06-30          20140630233203   
213432  rg7ayg1tob           2014-06-30          20140630233224   
213433  2f24umzkuv           2014-06-30          20140630233427   
213434  or77n2ojuj           2014-06-30          20140630233640   
213435  0a5bnb9bs4           2014-06-30          20140630233851   
213436  6fzrn49sfn           2014-06-30          20140630234113   
213437  r0jq0devgy           2014-06-30          20140630234243   
213438  v5lq9bj8gv           2014-06-30          20140630234429   
213439  msucfwmlzc           2014-06-30          20140630234729   
213440  04y8115avm           2014-06-30          20140630234933   
213441  omlc9iku7t           2014-06-30          20140630235151   
213442  rf0ay567js           2014-06-30          20140630235309   
213443  0k26r3mir0           2014-06-30          20140630235340   
213444  40o1ivh6cb           2014-06-30          20140630235352   
213445  qbxza0xojf           2014-06-30          20140630235547   
213446  zxodksqpep           2014-06-30          20140630235636   
213447  mhewnxesx9           2014-06-30          20140630235719   
213448  6o3arsjbb4           2014-06-30          20140630235754

In [14]:
sessions

user_id                 action       action_type  \
0         d1mm9tcy42                 lookup               NaN   
1         d1mm9tcy42         search_results             click   
2         d1mm9tcy42                 lookup               NaN   
3         d1mm9tcy42         search_results             click   
4         d1mm9tcy42                 lookup               NaN   
5         d1mm9tcy42         search_results             click   
6         d1mm9tcy42                 lookup               NaN   
7         d1mm9tcy42            personalize              data   
8         d1mm9tcy42                  index              view   
9         d1mm9tcy42                 lookup               NaN   
10        d1mm9tcy42         search_results             click   
11        d1mm9tcy42                 lookup               NaN   
12        d1mm9tcy42            personalize              data   
13        d1mm9tcy42                  index              view   
14        d1mm9tcy42       similar_listings              data   
15        d1mm9tcy42  ajax_refresh_subtotal             click   
16        d1mm9tcy42       similar_listings              data   
17        d1mm9tcy42  ajax_refresh_subtotal             click   
18        d1mm9tcy42                   show               NaN   
19        d1mm9tcy42            personalize              data   
20        d1mm9tcy42                   show              view   
21        d1mm9tcy42                 lookup               NaN   
22        d1mm9tcy42         search_results             click   
23        d1mm9tcy42                 lookup               NaN   
24        d1mm9tcy42         search_results             click   
25        d1mm9tcy42            personalize              data   
26        d1mm9tcy42                 lookup               NaN   
27        d1mm9tcy42                  index              view   
28        d1mm9tcy42                   show               NaN   
29        d1mm9tcy42            personalize              data   
...              ...                    ...               ...   
10567707  87k0fy4ugm                 update            submit   
10567708  87k0fy4ugm                 create         -unknown-   
10567709  87k0fy4ugm            personalize              data   
10567710  87k0fy4ugm         header_userpic              data   
10567711  87k0fy4ugm                 create            submit   
10567712  87k0fy4ugm                   show              view   
10567713  87k0fy4ugm         header_userpic              data   
10567714  87k0fy4ugm            personalize              data   
10567715  87k0fy4ugm  ajax_refresh_subtotal             click   
10567716  87k0fy4ugm       similar_listings              data   
10567717  87k0fy4ugm                  index              view   
10567718  87k0fy4ugm          confirm_email             click   
10567719  87k0fy4ugm           authenticate            submit   
10567720  87k0fy4ugm              dashboard              view   
10567721  9uqfg8txu3                   show              view   
10567722  9uqfg8txu3                 update            submit   
10567723  9uqfg8txu3                   edit              view   
10567724  9uqfg8txu3  languages_multiselect         -unknown-   
10567725  9uqfg8txu3                 create            submit   
10567726  9uqfg8txu3               callback  partner_callback   
10567727  9uqfg8txu3                 active         -unknown-   
10567728  9uqfg8txu3                 create            submit   
10567729  9uqfg8txu3         header_userpic              data   
10567730  9uqfg8txu3                   edit              view   
10567731  9uqfg8txu3                   edit              view   
10567732  9uqfg8txu3              dashboard              view   
10567733  9uqfg8txu3                   edit              view   
10567734  9uqfg8txu3          webcam_upload         -unknown-   
10567735  9uqfg8txu3                 active         -unknown-   
10567736  9uqfg8txu3       show_personalize     

In [15]:
print train.describe()
print '-------------------------------'
print train['gender'].value_counts()
print '-------------------------------'
print train['signup_method'].value_counts()
print '-------------------------------'
print train['language'].value_counts()
print '-------------------------------'
print train['affiliate_channel'].value_counts()
print '-------------------------------'
print train['affiliate_provider'].value_counts()
print '-------------------------------'
print train['first_affiliate_tracked'].value_counts()
print '-------------------------------'
print train['signup_app'].value_counts()
print '-------------------------------'
print train['first_browser'].value_counts()
print '-------------------------------'
print train['country_destination'].value_counts()

       timestamp_first_active            age    signup_flow
count            2.134510e+05  125461.000000  213451.000000
mean             2.013085e+13      49.668335       3.267387
std              9.253717e+09     155.666612       7.637707
min              2.009032e+13       1.000000       0.000000
25%              2.012123e+13      28.000000       0.000000
50%              2.013091e+13      34.000000       0.000000
75%              2.014031e+13      43.000000       0.000000
max              2.014063e+13    2014.000000      25.000000
-------------------------------
-unknown-    95688
FEMALE       63041
MALE         54440
OTHER          282
Name: gender, dtype: int64
-------------------------------
basic       152897
facebook     60008
google         546
Name: signup_method, dtype: int64
-------------------------------
en    206314
zh      1632
fr      1172
es       915
ko       747
de       732
it       514
ru       389
pt       240
ja       225
sv       122
nl        97
tr        64
d

In [16]:
print sessions.describe()
print '-------------------------------'
print sessions['action'].value_counts()
print '-------------------------------'
print sessions['action_type'].value_counts()
print '-------------------------------'
print sessions['action_detail'].value_counts()
print '-------------------------------'

          secs_elapsed
count  10431706.000000
mean      19405.810751
std       88884.243208
min           0.000000
25%         229.000000
50%        1147.000000
75%        8444.000000
max     1799977.000000
-------------------------------
show                           2768278
index                           843699
search_results                  725226
personalize                     706824
search                          536057
ajax_refresh_subtotal           487744
update                          365130
similar_listings                364624
social_connections              339000
reviews                         320591
active                          188036
similar_listings_v2             168788
lookup                          162041
create                          155887
dashboard                       152952
header_userpic                  141830
collections                     124417
edit                            109083
campaigns                       105028
track_page_view     

The goal here is to incorporate both sessions and users data into one dataset. There are several obstacles:

* <b>Age</b> column has various outliers e.g 2014, 1, 1956, etc. Solution: leave only valid values and categorize the other into '0-14', '100-1900', '1900-2014'
* <b>Dates</b> are needed to be broken into year, month, day, season, weekday, difference and so on.
* Categorical data should be one-hot encoded, to reduce the feature space mostly only 5-10 most common categories considered
* In sessions we need to group everything by id, aggregating with different functions to produce new features: number of actions, common unique actions, mean time elapsed, device types and so on
* NaNs. In sessions we just do not count it. In users fill them with -1

In [17]:
action_count = groupby.count()['action']
action_unique = groupby['action'].unique()
action_type_unique = groupby['action_type'].unique()
action_detail_unique = groupby['action_detail'].unique()
device_type_unique = groupby['device_type'].unique()

for key in action_unique.keys():
    action_unique[key] = len(action_unique[key])
    action_type_unique[key] = len(action_type_unique[key])
    action_detail_unique[key] = len(action_detail_unique[key])
    device_type_unique[key] = len(device_type_unique[key])

In [18]:
devices = ['Mac Desktop', 'Windows Desktop', 'iPhone', 'Android Phone', 'iPad Tablet', 'Other']
actions = ['show', 'index', 'search_results', 'personalize', 'search', 'ajax_refresh_subtotal', 'update', 'similar_listings', 
           'social_connections', 'reviews', 'other']
actions_type = ['view', 'data', 'click', '-unknown-', 'submit', 'message_post','partner_callback', 'booking_request', 'modify',
                'booking_response']

In [19]:
d_devices = {}
d_actions = {}
d_actions_type = {}
count =0
for g in groupby:
    top = g[1]['action'].value_counts().keys()
    common = set(top[:min(10,len(top))])&set(actions)
    bin0 = []
    for action in actions:
        if action in common:
            bin0.append(1)
        else:
            bin0.append(0)
    d_actions[g[0]] = bin0
    #print actions_bin
    top = g[1]['action_type'].value_counts().keys()
    common = set(top[:min(10,len(top))])&set(actions_type)
    bin0 = []
    for action in actions_type:
        if action in common:
            bin0.append(1)
        else:
            bin0.append(0)
    d_actions_type[g[0]] = bin0
    top = g[1]['device_type'].value_counts().keys()
    common = set(top[:min(6,len(top))])&set(devices)
    bin0 = []
    for device in devices:
        if device in common:
            bin0.append(1)
        else:
            bin0.append(0)
    d_devices[g[0]] = bin0

In [20]:
df_devices = pd.DataFrame(columns=devices)
df_devices = df_devices.from_dict(d_devices, orient='index')
df_devices.columns = devices
df_actions_type = pd.DataFrame(columns=actions_type)
df_actions_type = df_actions_type.from_dict(d_actions_type, orient='index')
df_actions_type.columns = actions_type
df_actions = pd.DataFrame(columns=actions)
df_actions = df_actions.from_dict(d_actions, orient='index')
df_actions.columns = actions

In [21]:
secs_elapsed_sum = groupby['secs_elapsed'].sum()
secs_elapsed_mean = groupby['secs_elapsed'].mean()
secs_elapsed_min = groupby['secs_elapsed'].min()
secs_elapsed_max = groupby['secs_elapsed'].max()

In [22]:
df_actions['secs_elapsed_sum'] = secs_elapsed_sum
df_actions['secs_elapsed_min'] = secs_elapsed_min
df_actions['secs_elapsed_max'] = secs_elapsed_max
df_actions['secs_elapsed_mean'] = secs_elapsed_mean 
df_actions['device_type_unique'] = device_type_unique
df_actions['action_type_unique'] = action_type_unique
df_actions['action_detail_unique'] = action_detail_unique
df_actions['action_unique'] = action_unique
df_actions['action_count'] = action_count
df_actions['id'] = df_actions.index
df_actions_type['id'] = df_actions_type.index
df_devices['id'] = df_devices.index

In [23]:
df_sessions = pd.merge(df_devices, df_actions, on='id')
df_sessions = pd.merge(df_sessions, df_actions_type, on='id')

In [24]:
df_sessions.to_csv('sessions_comp.csv')

In [25]:
date_account_created = train['date_account_created']
timestamp_first_active = train['timestamp_first_active']
ids = train['id']

In [26]:
import datetime
import calendar
dates = {}
dates_header = ['id','acc_year','acc_month','acc_day','stamp_year','stamp_month','stamp_day','stamp_hour','acc_weekday',
                'stamp_weekday','acc_season','stamp_season','date_diff']
seasons = ['winter', 'winter', 'spring', 'spring', 'spring', 'summer', 'summer', 'summer', 'autumn', 'autumn', 'autumn', 'winter']
for key in timestamp_first_active.keys():
    acc_year = int(date_account_created[key][0:4])
    acc_month = int(date_account_created[key][5:7])
    acc_day = int(date_account_created[key][8:10])
    stamp_year = int(str(timestamp_first_active[key])[0:4])
    stamp_month = int(str(timestamp_first_active[key])[4:6])
    stamp_day = int(str(timestamp_first_active[key])[6:8])
    stamp_hour = int(str(timestamp_first_active[key])[8:10])
    date1 = datetime.date(acc_year, acc_month, acc_day)
    acc_weekday = calendar.day_name[date1.weekday()]
    date2 = datetime.date(stamp_year, stamp_month, stamp_day)
    stamp_weekday = calendar.day_name[date2.weekday()]
    acc_season = seasons[acc_month-1]
    stamp_season = seasons[stamp_month-1]
    date_diff = (date1-date2).days
    dates[key] = [ids[key],acc_year,acc_month,acc_day,stamp_year,stamp_month,stamp_day,stamp_hour,acc_weekday,stamp_weekday,
                  acc_season,stamp_season,date_diff]
df_dates = pd.DataFrame(columns=dates_header)
df_dates = df_dates.from_dict(dates, orient='index')
df_dates.columns = dates_header

In [27]:
ages = train['age']

In [28]:
d_ages = {}
ages_header = ['id', 'normal_age', '0-14', '100-1900', '1900-2014']
for key in ages.keys():
    arr = [0]*5
    if not pd.isnull(ages[key]):
        if ages[key]>=1900:
            arr[4] = 1
        if (ages[key]<1900) and (ages[key]>=100):
            arr[3] = 1  
        if (ages[key]<14) and (ages[key]>=0):
            arr[2] = 1 
        if (ages[key]<100) and (ages[key]>=14):
            arr[1] = ages[key] 
    else:
        arr = [-1]*5
    arr[0] = ids[key]
    d_ages[key] = arr
df_ages = pd.DataFrame(columns=ages_header)
df_ages = df_ages.from_dict(d_ages, orient='index')
df_ages.columns = ages_header

In [29]:
browser = []
allowed = set(['Chrome','Safari','Firefox','-unknown-','IE','Mobile Safari'])
for v in train['first_browser']:
    if v in allowed:
        browser.append(v)
    else:
        browser.append('other')

In [30]:
train['first_browser'] = browser

In [31]:
del train['date_account_created']
del train['timestamp_first_active']
del train['date_first_booking']
del train['age']
del train['first_device_type']

In [32]:
df_users = pd.merge(train, df_dates, on='id')
df_users = pd.merge(df_users, df_ages, on='id')
df_users.to_csv('train_users.csv')

In [36]:
train_preprocessed = pd.merge(df_users, df_sessions, on='id')
train_preprocessed.shape

(73815, 63)

After all, we left with the nice dataset with 73815 objects and 63 features (before one-hot encoding) with a small amount of NaNs. Features are: 

In [63]:
for f in train_preprocessed.columns:
    print f

id
gender
signup_method
signup_flow
language
affiliate_channel
affiliate_provider
first_affiliate_tracked
signup_app
first_browser
country_destination
acc_year
acc_month
acc_day
stamp_year
stamp_month
stamp_day
stamp_hour
acc_weekday
stamp_weekday
acc_season
stamp_season
date_diff
normal_age
0-14
100-1900
1900-2014
Mac Desktop
Windows Desktop
iPhone
Android Phone
iPad Tablet
Other
show
index
search_results
personalize
search
ajax_refresh_subtotal
update
similar_listings
social_connections
reviews
other
secs_elapsed_sum
secs_elapsed_min
secs_elapsed_max
secs_elapsed_mean
device_type_unique
action_type_unique
action_detail_unique
action_unique
action_count
view
data
click
-unknown-
submit
message_post
partner_callback
booking_request
modify
booking_response


In [37]:
len(train_preprocessed[train_preprocessed.isnull().any(axis=1)])

1408

In [35]:
train_preprocessed = train_preprocessed.fillna(-1)
train_preprocessed.to_csv('train_preprocessed.csv')

In [38]:
train = pd.read_csv('train_preprocessed.csv')
del train['Unnamed: 0']
del train['id']
y = train['country_destination'].copy()
del train['country_destination']
X = train.copy()
X = pd.get_dummies(X)
print X.shape

(73815, 144)


Now we just run k-means clustering to find out if there is a strict border between classes.

In [39]:
from sklearn.cluster import KMeans
est = KMeans(n_clusters=12)
est.fit(X)
labels = est.labels_

In [49]:
pd.Series(zip(labels, y)).value_counts()

(6, NDF)       17340
(0, NDF)        9320
(5, NDF)        6307
(6, US)         4730
(0, US)         4288
(5, US)         3271
(2, NDF)        3138
(11, NDF)       2794
(7, NDF)        2675
(2, US)         1877
(11, US)        1759
(9, NDF)        1674
(7, US)         1353
(9, US)         1307
(4, NDF)        1008
(6, other)       864
(4, US)          772
(0, other)       738
(5, other)       576
(1, NDF)         501
(1, US)          451
(11, other)      354
(2, other)       345
(6, FR)          310
(0, FR)          274
(5, FR)          271
(7, other)       260
(9, other)       241
(3, NDF)         213
(3, US)          213
               ...  
(7, PT)            9
(11, PT)           8
(1, CA)            7
(10, FR)           7
(3, ES)            6
(1, DE)            6
(3, NL)            6
(3, CA)            6
(3, GB)            6
(9, PT)            5
(8, NDF)           5
(4, AU)            5
(10, IT)           4
(1, NL)            4
(10, GB)           3
(3, DE)            3
(10, ES)     

In [50]:
pd.Series(labels, y).value_counts()


6     24000
0     15365
5     11018
2      5858
11     5377
7      4622
9      3558
4      2151
1      1163
3       524
10      169
8        10
dtype: int64

Well, there is not